File: 1992-find-all-groups-of-farmland/1992-find-all-groups-of-farmland.java
class Solution {
    public int[][] findFarmland(int[][] land) {
        int rows = land.length;
        int cols = land[0].length;
        
        List<int[]> list = new ArrayList<>();
        for(int i=0; i<rows; i++) {
            for(int j=0; j<cols; j++) {
                if(land[i][j] == 1) {
                    extendReach(i, j, land, list);
                }
            }
        }
        return list.toArray(new int[list.size()][]);
    }
    
    private void extendReach(int x, int y, int[][] land, List<int[]> list) {
        int endX = x;
        int endY = y;
        //Find ending y co-ordinate for the Farmland
        while(endY+1 < land[0].length && land[x][endY+1] == 1) {
            land[x][endY+1] = 0;
            endY++;
        }
        
        //Find ending x co-ordinate for the Farmland
        while(endX+1<land.length && land[endX+1][y] == 1) {
            land[endX+1][y] = 0;
            endX++;
        }
        
        //Fill the rectangular Farmland with zeroes
        //To avoid revisitng/recounting
        for(int i=x; i<=endX; i++) {
            for(int j=y; j<=endY; j++) {
                land[i][j] = 0;
            }
        }
        
        list.add(new int[]{x, y, endX, endY});
    }
}

File: 1992-find-all-groups-of-farmland/1992-find-all-groups-of-farmland.cpp
class Solution {
public:
    vector<vector<int>> findFarmland(vector<vector<int>>& land) {
        m = land.size();
        n = land[0].size();
        vector<vector<int>> v;
        for(int i=0; i<m; ++i)
        {
            for(int j=0; j<n; ++j)
            {
                if(land[i][j] == 1)
                {
                    int endi=i, endj=j;
                    dfs(land, i, j, endi, endj);
                    v.push_back({i, j, endi, endj});
                }
            }
        }
        return v;
    }
private:
    void dfs(vector<vector<int>>& land, int i, int j, int& endi, int& endj)
    {
        land[i][j] = -1;
        endi = max(i, endi);
        endj = max(j, endj);
        if(i+1 < m && land[i+1][j] == 1)
        {
            dfs(land, i+1, j, endi, endj);
        }
        if(j+1 < n && land[i][j+1] == 1)
        {
            dfs(land, i, j+1, endi, endj);
        }
    }
    int m;
    int n;
};